Makefile.am

```diff
@@ -63,11 +63,64 @@ AM_CPPFLAGS = \
     $(ICU_CFLAGS)
 
 if HAVE_LDAP
-AM_CPPFLAGS += $(LDAP_CPPFLAGS)
+    AM_CPPFLAGS += $(LDAP_CPPFLAGS)
 endif # HAVE_LDAP
 
 AM_LDFLAGS = $(COV_LDFLAGS) $(ICU_LIBS)
 
+if ASAN
+    LSAN_SUPPRESS="env LSAN_OPTIONS=suppressions=leaksanitizer.suppress"
+    AM_CFLAGS += -fsanitize=address -DWITH_ASAN
+    AM_CXXFLAGS += -fsanitize=address -DWITH_ASAN
+    AM_LDFLAGS += -fsanitize=address
+
+if STATIC_ASAN
+        AM_CFLAGS += -DWITH_STATIC_ASAN
+        AM_CXXFLAGS += -DWITH_STATIC_ASAN
+
+if IS_GCC_COMPILER
+            AM_CFLAGS += -static-libasan -DWITH_GCC
+            AM_CXXFLAGS += -static-libasan -DWITH_GCC
+            AM_LDFLAGS += -static-libasan
+endif # IS_GCC_COMPILER
+else # ! STATIC_ASAN
+if IS_GCC_COMPILER
+            AM_CFLAGS += -DWITH_GCC
+            AM_CXXFLAGS += -DWITH_GCC
+else # ! IS_GCC_COMPILER
+            AM_CFLAGS += -shared-libsan
+            AM_CXXFLAGS += -shared-libsan
+            AM_LDFLAGS += -shared-libsan
+endif # ! IS_GCC_COMPILER
+endif # STATIC_ASAN
+endif # ASAN
+
+if UBSAN
+    AM_CFLAGS += -fsanitize=undefined -DWITH_UBSAN
+    AM_CXXFLAGS += -fsanitize=undefined -DWITH_UBSAN
```

Comment:
I feel like we could maybe have done all this decision-making in configure.ac, setting up SAN_CFLAGS, SAN_CXXFLAGS, SAN_LDFLAGS with the appropriate combinations of flags (or empty if both sanitisers are disabled), and then in Makefile.am we would just need to interpolate these new variables into their respective AM_foo once, unconditionally.  We do something like this (albeit simpler) for GCOV_CFLAGS et al, so have a look at that for an example.

Off the top of my head, I'm not sure if it would be more readable/maintainable, but I notice we do already have some (all?) of these conditions in configure.ac already, e.g. for setting asan_desc etc.  If we're gonna do it all in there anyway, might as well not redo it later.